dandi version error in staging

```pytb
def test_automatic_dandi_upload(self):
>       automatic_dandi_upload(dandiset_id="200560", nwb_folder_path=self.nwb_folder_path, staging=True)

tests/test_minimal/test_tools/test_dandi_transfer_tools.py:45: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/opt/hostedtoolcache/Python/3.8.16/x64/lib/python3.8/site-packages/neuroconv/tools/data_transfers.py:[36](https://github.com/catalystneuro/neuroconv/actions/runs/4378541440/jobs/7663420144#step:9:37)4: in automatic_dandi_upload
    dandi_upload(paths=[str(x) for x in organized_nwbfiles], dandi_instance=dandi_instance)
/opt/hostedtoolcache/Python/3.8.16/x64/lib/python3.8/site-packages/dandi/upload.py:70: in upload
    client.check_schema_version()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <dandi.dandiapi.DandiAPIClient object at 0x7f3f9cb34910>
schema_version = '0.6.3'

    def check_schema_version(self, schema_version: Optional[str] = None) -> None:
        """
        Confirms that the server is using the same version of the Dandi schema
        as the client.  If it is not, a `SchemaVersionError` is raised.
    
        :param schema_version: the schema version to confirm that the server
            uses; if not set, the schema version for the installed
            ``dandischema`` library is used
        """
        if schema_version is None:
            schema_version = models.get_schema_version()
        server_info = self.get("/info/")
        server_schema_version = server_info.get("schema_version")
        if not server_schema_version:
            raise RuntimeError(
                "Server did not provide schema_version in /info/;"
                f" returned {server_info!r}"
            )
        if server_schema_version != schema_version:
>           raise SchemaVersionError(
                f"Server requires schema version {server_schema_version};"
                f" client only supports {schema_version}.  You may need to"
                " upgrade dandi and/or dandischema."
            )
E           dandi.exceptions.SchemaVersionError: Server requires schema version 0.6.4; client only supports 0.6.3.  You may need to upgrade dandi and/or dandischema.
```

--- a/dandi/download.py
+++ b/dandi/download.py
@@ -53,7 +53,7 @@ from .files import LocalAsset, find_dandi_files
 from .support.digests import get_digest, get_zarr_checksum
 from .support.iterators import IteratorWithAggregation
 from .support.pyout import naturalsize
-from .support.typing import Literal
+from .support.typing import Literal, Protocol
 from .utils import (
     abbrev_prompt,
     ensure_datetime,
@@ -520,6 +520,14 @@ def _populate_dandiset_yaml(
     }
 
 
+class Hasher(Protocol):
+    def update(self, data: bytes) -> None:
+        ...
+
+    def hexdigest(self) -> str:
+        ...
+
+
 def _download_file(
     downloader: Callable[[int], Iterator[bytes]],
     path: Path,
@@ -646,20 +654,26 @@ def _download_file(
 
     yield {"status": "downloading"}
 
-    algo, digester, digest, downloaded_digest = None, None, None, None
+    algo: Optional[str] = None
+    digester: Optional[Callable[[], Hasher]] = None
+    digest: Optional[str] = None
+    downloaded_digest: Optional[Hasher] = None
     if digests:
         # choose first available for now.
         # TODO: reuse that sorting based on speed
         for algo, digest in digests.items():
-            if algo == "dandi-etag":
+            if algo == "dandi-etag" and size is not None:
                 from dandischema.digests.dandietag import ETagHashlike
 
-                digester = lambda: ETagHashlike(size)  # noqa: E731
+                # Instantiate outside the lambda so that mypy is assured that
+                # `size` is not None:
+                hasher = ETagHashlike(size)
+                digester = lambda: hasher  # noqa: E731
             else:
                 digester = getattr(hashlib, algo, None)
-            if digester:
+            if digester is not None:
                 break
-        if not digester:
+        if digester is None:
             lgr.warning("Found no digests in hashlib for any of %s", str(digests))
 
     # TODO: how do we discover the total size????
@@ -725,12 +739,12 @@ def _download_file(
 
     if downloaded_digest and not resuming:
         assert downloaded_digest is not None
-        downloaded_digest = downloaded_digest.hexdigest()  # we care only about hex
+        final_digest = downloaded_digest.hexdigest()  # we care only about hex
         if digest_callback is not None:
             assert isinstance(algo, str)
-            digest_callback(algo, downloaded_digest)
-        if digest != downloaded_digest:
-            msg = f"{algo}: downloaded {downloaded_digest} != {digest}"
+            digest_callback(algo, final_digest)
+        if digest != final_digest:
+            msg = f"{algo}: downloaded {final_digest} != {digest}"
             yield {"checksum": "differs", "status": "error", "message": msg}
             lgr.debug("%s is different: %s.", path, msg)
             return

--- a/dandi/files/bids.py
+++ b/dandi/files/bids.py
@@ -92,7 +92,9 @@ class BIDSDatasetDescriptionAsset(LocalFileAsset):
                 self._asset_errors: dict[str, list[ValidationResult]] = defaultdict(
                     list
                 )
-                self._asset_metadata = defaultdict(BareAsset.unvalidated)
+                # Don't apply eta-reduction to the lambda, as mypy needs to be
+                # assured that defaultdict's argument takes no parameters.
+                self._asset_metadata = defaultdict(lambda: BareAsset.unvalidated())
                 for result in results:
                     if result.id in BIDS_ASSET_ERRORS:
                         assert result.path

--- a/dandi/metadata.py
+++ b/dandi/metadata.py
@@ -24,12 +24,13 @@ from uuid import uuid4
 from xml.dom.minidom import parseString
 
 from dandischema import models
+from pydantic import ByteSize, parse_obj_as
 import requests
 import tenacity
 
 from . import __version__, get_logger
 from .consts import metadata_all_fields
-from .misctypes import Digest, LocalReadableFile, Readable
+from .misctypes import DUMMY_DANDI_ETAG, Digest, LocalReadableFile, Readable
 from .pynwb_utils import (
     _get_pynwb_metadata,
     get_neurodata_types,
@@ -90,20 +91,21 @@ def get_metadata(
                 dandiset_path,
                 bids_dataset_description=bids_dataset_description,
             )
-            if not digest:
-                _digest = "0" * 32 + "-1"
-                digest = Digest.dandi_etag(_digest)
-            path_metadata = df.get_metadata(digest=digest)
             assert isinstance(df, bids.BIDSAsset)
+            if not digest:
+                digest = DUMMY_DANDI_ETAG
+            path_metadata = df.get_metadata(digest=digest)
             meta["bids_version"] = df.get_validation_bids_version()
             # there might be a more elegant way to do this:
-            for key in metadata_all_fields:
-                try:
-                    value = getattr(path_metadata.wasAttributedTo[0], key)
-                except AttributeError:
-                    pass
-                else:
-                    meta[key] = value
+            if path_metadata.wasAttributedTo is not None:
+                attributed = path_metadata.wasAttributedTo[0]
+                for key in metadata_all_fields:
+                    try:
+                        value = getattr(attributed, key)
+                    except AttributeError:
+                        pass
+                    else:
+                        meta[key] = value
 
     if r.get_filename().endswith((".NWB", ".nwb")):
         if nwb_has_external_links(r):
@@ -1002,13 +1004,14 @@ def add_common_metadata(
         metadata.blobDateModified = mtime
         if mtime > metadata.dateModified:
             lgr.warning("mtime %s of %s is in the future", mtime, r)
-    metadata.contentSize = r.get_size()
+    size = r.get_size()
     if digest is not None and digest.algorithm is models.DigestType.dandi_zarr_checksum:
         m = re.fullmatch(
             r"(?P<hash>[0-9a-f]{32})-(?P<files>[0-9]+)--(?P<size>[0-9]+)", digest.value
         )
         if m:
-            metadata.contentSize = int(m["size"])
+            size = int(m["size"])
+    metadata.contentSize = parse_obj_as(ByteSize, size)
     if metadata.wasGeneratedBy is None:
         metadata.wasGeneratedBy = []
     metadata.wasGeneratedBy.append(get_generator(start_time, end_time))

--- a/dandi/metadata.py
+++ b/dandi/metadata.py
@@ -18,7 +18,6 @@ from typing import (
     Type,
     TypeVar,
     Union,
-    cast,
 )
 from uuid import uuid4
 from xml.dom.minidom import parseString
@@ -625,7 +624,7 @@ M = TypeVar("M", bound=models.DandiBaseModel)
 
 
 def extract_model(modelcls: Type[M], metadata: dict, **kwargs: Any) -> M:
-    m = cast(M, modelcls.unvalidated())
+    m = modelcls.unvalidated()
     for field in m.__fields__.keys():
         value = kwargs.get(field, extract_field(field, metadata))
         if value is not None:

--- a/dandi/support/typing.py
+++ b/dandi/support/typing.py
@@ -1,6 +1,6 @@
 import sys
 
 if sys.version_info >= (3, 8):
-    from typing import Literal, TypedDict  # noqa: F401
+    from typing import Literal, Protocol, TypedDict  # noqa: F401
 else:
-    from typing_extensions import Literal, TypedDict  # noqa: F401
+    from typing_extensions import Literal, Protocol, TypedDict  # noqa: F401

--- a/dandi/tests/test_upload.py
+++ b/dandi/tests/test_upload.py
@@ -225,6 +225,7 @@ def test_upload_bids_metadata(
         if "sub-" in apath:
             metadata = dandiset.get_asset_by_path(apath).get_metadata()
             # Hard-coded check for the subject identifier set in the fixture:
+            assert metadata.wasAttributedTo is not None
             assert metadata.wasAttributedTo[0].identifier == "Sub1"
 
 

--- a/setup.cfg
+++ b/setup.cfg
@@ -34,7 +34,7 @@ install_requires =
     bidsschematools >= 0.5.0
     click
     click-didyoumean
-    dandischema ~= 0.7.0
+    dandischema ~= 0.8.0
     etelemetry >= 0.2.2
     fasteners
     fscacher >= 0.3.0

